tests/abi/tools/revng-abi-verify/ABIArtifactParser.cpp

```diff
@@ -0,0 +1,188 @@
+/// \file Verify.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <algorithm>
+
+#include "revng/ABI/FunctionType.h"
+#include "revng/Model/Binary.h"
+
+#include "ABIArtifactParser.h"
+
+namespace State {
+
+using namespace abi::artifact;
+
+struct SingleRun {
+  llvm::StringRef Function;
+  bool IsLittleEndian;
+  Registers Registers;
+  Stack Stack;
+  Arguments ReturnValue;
+  Arguments Arguments;
+};
+using Deserialized = llvm::SmallVector<SingleRun, 16>;
+
+} // namespace State
+
+//
+// `llvm::yaml` traits for the internal state.
+//
+
+template<>
+struct llvm::yaml::MappingTraits<State::Register> {
+  static void mapping(IO &IO, State::Register &N) {
+    IO.mapRequired("Name", N.Name);
+    IO.mapRequired("Value", N.Value);
+  }
+};
+
+template<>
+struct llvm::yaml::MappingTraits<State::StackValue> {
+  static void mapping(IO &IO, State::StackValue &N) {
+    IO.mapRequired("Offset", N.Offset);
+    IO.mapRequired("Value", N.Value);
+  }
+};
+
+template<>
+struct llvm::yaml::MappingTraits<State::Argument> {
+  static void mapping(IO &IO, State::Argument &N) {
+    IO.mapRequired("Type", N.Type);
+    IO.mapRequired("Value", N.Value);
+    IO.mapRequired("Address", N.Address);
+  }
+};
+
+template<>
+struct llvm::yaml::MappingTraits<State::SingleRun> {
+  static void mapping(IO &IO, State::SingleRun &N) {
+    IO.mapRequired("Function", N.Function);
+    IO.mapRequired("IsLittleEndian", N.IsLittleEndian);
+    IO.mapRequired("Registers", N.Registers);
+    IO.mapRequired("Stack", N.Stack);
+    IO.mapRequired("Arguments", N.Arguments);
+    IO.mapRequired("ReturnValue", N.ReturnValue);
+  }
+};
+
+template<typename Type>
+struct GenericSmallVectorYamlTrait {
+  static size_t size(llvm::yaml::IO &, Type &Value) { return Value.size(); }
+  static auto &element(llvm::yaml::IO &, Type &Value, size_t Index) {
+    if (Index >= Value.size())
+      Value.resize(Index + 1);
+    return Value[Index];
+  }
+};
+
+template<>
+struct llvm::yaml::SequenceTraits<State::Registers>
+  : GenericSmallVectorYamlTrait<State::Registers> {};
+template<>
+struct llvm::yaml::SequenceTraits<State::Stack>
+  : GenericSmallVectorYamlTrait<State::Stack> {};
+template<>
+struct llvm::yaml::SequenceTraits<State::Arguments>
+  : GenericSmallVectorYamlTrait<State::Arguments> {};
+template<>
+struct llvm::yaml::SequenceTraits<State::Deserialized>
+  : GenericSmallVectorYamlTrait<State::Deserialized> {};
```

Comment:
I had no idea those were defined already. I guess it's safe to drop them.

Reply:
On the other hand, I feel like "give me the default `SmallVector` sequence traits" is more explicit than defining `SequenceElementTraits` for the same type. And since a three-line specialization is required in either case, does it really matter if I just choose the more explicit one?

Reply:
If the body ever changes on the LLVM side to some internal changes, we'll have to update it.
Let's stick with default.

Reply:
Done: https://github.com/revng/revng/pull/148/commits/cd1439dfa5a530eedd9c0fbaa674f2406750e83a.